### PR TITLE
[ADDED] Distributed Message Tracing

### PR DESCRIPTION
**Pull request by kozlovic**
When an incoming message contains the header `Nats-Trace-Dest` set to a valid subject, the server will send there messages representing events happening in each server where the message is processed.

The events will give details about ingress, subject mapping, stream export and service imports, jetstream and egress to subscriptions and/or routes, leafnodes and gateways. Except for a standalone server it is expected to receive multiple trace messages for a given inbound message.

The header `Nats-Trace-Only`, if set to `true`, will produce the same tracing events than without that header, except that the message will not be delivered to subscriptions or inserted in the JetStream stream. This is usefull to see the path that a message would take but without affecting running applications.

Note that in a mixed environment where not all servers would be running at 2.11+, when a message is supposed to be traced only and the remote does not understand message tracing, the message will not be forwared and the Egress event for that remote will indicate the reason.

Signed-off-by: Ivan Kozlovic <ivan@synadia.com>

### CODE REVIEW

**Comment by ripienaar** [comment]
@kozlovic  When tracing over an account import/export is it expected that I wont get trace messages once the message cross the account boundary?

Client on account "one" in london -> imports weather.> from account weather -> client on account weather connected across a gateway

In this case I get only 1 trace event with 

```
  "events": [
    {
      "type": "in",
      "kind": 0,
      "cid": 859,
      "name": "NATS CLI Version development",
      "acc": "one",
      "subj": "weather.lon"
    },
    {
      "type": "si",
      "acc": "weather",
      "from": "weather.lon",
      "to": "weather.lon"
    }
  ]
```

In the same account I get events from otherside of the gateway also

I guess it could be a security concern to trace it through but otoh when debugging would be great to have

**Comment by kozlovic** [comment]
@ripienaar The decision was to go through service import only if "share: true" is specified in the service import. This was for the reason you just mentioned: security concern. If you have sharing enabled in the service import, then there may be an issue. I have test for service import with sharing/not sharing across super cluster and it passes, but there may still be an issue.

**Comment by ripienaar** [comment]
> @ripienaar The decision was to go through service import only if "share: true" is specified in the service import. This was for the reason you just mentioned: security concern. If you have sharing enabled in the service import, then there may be an issue. I have test for service import with sharing/not sharing across super cluster and it passes, but there may still be an issue.

I didnt have share on, will try

**Comment by kozlovic** [comment]
@ripienaar I am thinking of adding something to the MsgTraceEgress structure to be able to "link" MsgTraceEvent's from a network topology together. Something like this:
```
diff --git a/server/msgtrace.go b/server/msgtrace.go
index 88c35d5f..601cf16d 100644
--- a/server/msgtrace.go
+++ b/server/msgtrace.go
@@ -110,6 +110,11 @@ type MsgTraceEgress struct {
        Subscription string `json:"sub,omitempty"`
        Queue        string `json:"queue,omitempty"`
        Error        string `json:"error,omitempty"`
+
+       // This is for applications that unmarshal the trace events
+       // and want to link an egress to route/leaf/gateway with
+       // the MsgTraceEvent from that server.
+       Link *MsgTraceEvent `json:"-"`
 }
 
 // -------------------------------------------------------------
```
The idea is that the library that assembles message trace events would then return only one event (the one that had a CLIENT as an Ingress) and then by going through the Events list, when getting an Egress that is not a CLIENT kind, there would be the Link field that would not be nil and you could descend the graph. If the Link is nil, it would be an indication that this trace from that server (as given by the Egress's info) is missing.

Any thoughts? Would you think that it should be done differently?

**Comment by ripienaar** [comment]
> Any thoughts? Would you think that it should be done differently?


Yeah anything you can do to make assembling the flows easier would help (been trying to hack up a sorter/linker for them and its difficult!)

**Comment by kozlovic** [comment]
@ripienaar I will have the library repo ready today.

**Comment by kozlovic** [comment]
@ripienaar I am coming to the realization that I don't think we need the "Hop" because the Ingress/Egress has already the name of the server the message comes from/sent to, so that should be enough to link them. As for the Hops (the count at the MsgTraceEvent level), again, it is nice, but it is basically the number of Egress that are not a CLIENT kind, so again, not strictly necessary.

**Comment by ripienaar** [comment]
> @ripienaar I am coming to the realization that I don't think we need the "Hop" because the Ingress/Egress has already the name of the server the message comes from/sent to, so that should be enough to link them. As for the Hops (the count at the MsgTraceEvent level), again, it is nice, but it is basically the number of Egress that are not a CLIENT kind, so again, not strictly necessary.

Yeah sounds reasonable :) 

Kind of this is why I wanted a kind of draft client to assemble them, we'll discover a lot in doing that

**Comment by MauriceVanVeen** [comment]
Really cool! Have been trying it out for a bit. :star_struck: 

Not sure if there's also an intention to add tracing between stream => consumer as well as mirroring/sourcing?
Am thinking of a use case where someone publishes into stream 1, a consumer consumes from that stream and publishes 1 or more events into stream 2.
Would that also become available with this tracing, or would that be an extension to this?

Thinking along the lines of a full graph, from the initial message all the way down to how it was processed. (From an OpenTelemetry perspective that could be one 'trace' consisting of multiple 'spans')

**Comment by kozlovic** [comment]
> Would that also become available with this tracing, or would that be an extension to this?

@MauriceVanVeen No, currently the only JetStream related tracing is to know if a message would have been inserted to a stream, with which subject, and if not what caused the message to not be inserted (say presence of a header that causes server checks to reject it).

**Comment by hwinkel** [comment]
How are the traces exported? Towards "standard" OTEL collectors like jaeger, tempo etc. ? Does the export happens on a per serves bases or centrally like surveyor does ?

**Comment by ripienaar** [comment]
> How are the traces exported? Towards "standard" OTEL collectors like jaeger, tempo etc. ? Does the export happens on a per serves bases or centrally like surveyor does ?

Now the messages are published as JSON (see `msgtrace.go` for the types) to keep the server light weight, I imagine you'd run a collector process next to your nats servers and the purpose of that would be to receive these, assemble them and send them to otel etc

For now otel/jaeger etc headers do not trigger a trace (though our service latency tracing does support this see https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-3.md) 

The focus of this is primarily debugging though and not, for example, tracing 100% of message routing

**Comment by bruth** [comment]
@hwinkel You can think of this as the lower layer of the output. It is published to the designated trace subject which then the subscribed client would do the work to map/export to otel (or anything else). That is the next bit of work to write the exporter.

**Comment by bruth** [comment]
And second what @ripienaar said. The scope was initially for trace debugging, albeit, depending on throughput/load of a system, you could opt to trace whatever you want.

**Comment by ripienaar** [comment]
For community who are following these are some renders of traces in various cases

https://gist.github.com/ripienaar/78666080a4296bc5a0e317d1c9c08df8

The cluster is a 3 cluster super cluster each with 3 nodes and a single leafnode hanging off the `lon` cluster.

**Comment by kozlovic** [comment]
@ripienaar Just a little comment, I think you should use `%q` for the strings so they are enclosed with double quotes, it would help for the reading.

**Comment by davedotdev** [comment]
I spent some (not negligible) time on this in the networking space working for a large vendor. Multiple approaches were taken, some out-of-band and some in-band, but the outcomes were the same; the desire to have tracing within an [OAM](https://en.wikipedia.org/wiki/Operations,_administration,_and_management) domain, for targeted packets or a sample set. A networked system has some configuration to make it part of an OAM domain and each system in that domain must be able to understand the requirements of being part of a tracing capable system, or pass it on without modification exactly like you've done here with the header pass-thru.

What I wanted to raise for debate, was the ability to turn have tracing domains for NATS servers, so that one or more server or group of servers could ignore the header. For instance, this could be useful in secure environments, where some groups can trace and others cannot, or in situations where visibility needs to be reduced. Initial suggestion would be to have a server tag like `"OAMDomain": "[`X`, `Y`, `Z`]"`.  Is it something that's been thought about? Would any of the commenters here find it useful? I can see a requirement for it and if it changes any design decisions being made, it's worth raising for future considerations.

A good use-case would be hub and spoke, where the spokes are under your control, but the hubs are not. Traffic would traverse the hub in either case, but the traces would show the hub ingress and egress points and the resulting trace would show the delay, but not reveal anything about the hub.

Nice work Ivan - this really is great (and I'm not griping - just thinking about wider use cases!!).

**Comment by ripienaar** [comment]
> A good use-case would be hub and spoke, where the spokes are under your control, but the hubs are not. Traffic would traverse the hub in either case, but the traces would show the hub ingress and egress points and the resulting trace would show the delay, but not reveal anything about the hub.

for now our only control here is when crossing account barriers if you dont set sharing true you wont get traces from the other side.  One can imagine spoke/leafnode on an account and access to it over exports/imports would control visibility

I dont want to over complicate things from the get-go so but this is a good point for future iterations.  I imagine we'd want to eventually support disabling tracing on an account level for sensitive cases.  Much can be learned from seeing which mappings apply etc

**Comment by ripienaar** [comment]
For me I am happy with where this is to merge it. We might iterate and once it’s in a bit look at how it would be to trigger it from headers from other systems like we support for service traces 

for now in the interest of getting it in peoples hands via nightly builds this works well. 

To be clear I didn’t code review I looked at functionality only

**Comment by bruth** [comment]
I agree with getting it merged, knowing it may need some iteration. What are your thoughts @Jarema?

Once merged, I want to tag a preview release with this feature so people can access a build to try it out.

File: server/client.go
Comment (derekcollison) [inline server/client.go:3423]: Why would we qualify this here?

File: server/client.go
Comment (derekcollison) [inline server/client.go:3479]: Same here, not sure why we are qualifying based on tracing contexts now.

File: server/client.go
Comment (derekcollison) [inline server/client.go:3517]: If function call not too heavy hit on performance (need to measure) we could add check for receiver being nil in addEvent functions to clean up a bit.

File: server/client.go
Comment (derekcollison) [inline server/client.go:4344]: As noted I messed this up, let's switch to allow_trace for service export and stream import.

File: server/client.go
Comment (derekcollison) [inline server/client.go:4384]: Will this possibly fire NoResponders? And if so, is that the desired outcome?

File: server/events.go
Comment (derekcollison) [inline server/events.go:650]: change interface{} to any, been slowly moving over to that.

File: server/jetstream_cluster.go
Comment (derekcollison) [inline server/jetstream_cluster.go:7665]: Just call `mset.getAndDeleteMsgTrace`?

File: server/msgtrace.go
Comment (derekcollison) [inline server/msgtrace.go:267]: This should be inlined by compiler, can we double check?

File: server/stream.go
Comment (derekcollison) [inline server/stream.go:3257]: Since we pass nil most of the time for last arg, keep same and have that call into new func that takes extra arg?

**Review by derekcollison — COMMENTED** [review]
In general LGTM - Some comments and questions.

**Comment by Jarema** [comment]
Yes, agree with merging and doing a preview release @bruth .
This will allow us to get more community feedback which could be invaluable for feature like this one.
The only concern is conflicts when cherry picking for patch releases, but that is unavoidable, as anyway this branch would need constant rebases and updates which is similar cost.
Let's merge and iterate.

**Comment by ripienaar** [comment]
We're working on a fix about import sharing, will look to merge after

File: server/client.go
Comment (kozlovic) [reply server/client.go]: Because we don't want to update that if we were doing a trace only. I have changed isMsgTraceEnabled to actually return if we are doing trace only (which then is false when mt is nil or when mt is not but we are doing message delivery). I also added a comment to make it clear as to why now.

File: server/client.go
Comment (kozlovic) [reply server/client.go]: Seems to be ok. I did the ok, I ran the whole bench publish tests, but since it will be more prevalent when there are subscriptions (multiple egresses), here is an excerpt:
```
                                      │  new1.txt   │             new2.txt              │
                                      │   sec/op    │   sec/op     vs base              │
Publish/MsgSz=0b/Subs=1xAsync-8         200.7n ± 1%   200.1n ± 1%       ~ (p=0.416 n=6)
Publish/MsgSz=0b/Subs=1xAsyncQueue-8    209.6n ± 0%   210.2n ± 0%       ~ (p=0.065 n=6)
Publish/MsgSz=0b/Subs=10xAsync-8        880.2n ± 6%   885.2n ± 6%       ~ (p=0.818 n=6)
Publish/MsgSz=0b/Subs=10xAsyncQueue-8   264.0n ± 4%   265.4n ± 2%       ~ (p=0.669 n=6)
geomean                                 314.4n        315.3n       +0.28%
```

File: server/client.go
Comment (kozlovic) [reply server/client.go:4382]: No, since this is an "icb", we always return `true` in deliverMsg().

File: server/events.go
Comment (kozlovic) [reply server/events.go]: Done

File: server/jetstream_cluster.go
Comment (kozlovic) [reply server/jetstream_cluster.go]: Done

File: server/msgtrace.go
Comment (kozlovic) [reply server/msgtrace.go:266]: It is. See the lack of "CALL" and instead the code from msgtrace.go embedded in the deliverMsg() function:
```
TEXT github.com/nats-io/nats-server/v2/server.(*client).deliverMsg(SB) /Users/ik/dev/go/src/github.com/kozlovic/nats-server/server/client.go
  client.go:3336        0x1003596a0             f9400b90                MOVD 16(R28), R16                                                                       
  client.go:3336        0x1003596a4             d10803f1                SUB $512, RSP, R17                                                                      
  client.go:3336        0x1003596a8             eb10023f                CMP R16, R17                                                                            
  client.go:3336        0x1003596ac             5400a289                BLS 1300(PC)                                                                            
  client.go:3336        0x1003596b0             d10a03f4                SUB $640, RSP, R20                                                                      
  client.go:3336        0x1003596b4             a93ffa9d                STP (R29, R30), -8(R20)                                                                 
  client.go:3336        0x1003596b8             9100029f                MOVD R20, RSP                                                                           
  client.go:3336        0x1003596bc             d10023fd                SUB $8, RSP, R29                                                                        
  client.go:3336        0x1003596c0             f9015be4                MOVD R4, 688(RSP)                                                                       
  client.go:3336        0x1003596c4             f90167e7                MOVD R7, 712(RSP)                                                                       
  client.go:3336        0x1003596c8             f90173ea                MOVD R10, 736(RSP)                                                                      
  client.go:3336        0x1003596cc             f9017fed                MOVD R13, 760(RSP)                                                                      
  client.go:3336        0x1003596d0             39018bff                MOVB ZR, 98(RSP)                                                                        
  msgtrace.go:268       0x1003596d4             f9410410                MOVD 520(R0), R16                                                                       
  msgtrace.go:269       0x1003596d8             b4000070                CBZ R16, 3(PC)                                                                          
  msgtrace.go:272       0x1003596dc             3941a211                MOVBU 104(R16), R17                                                                     
  client.go:3338        0x1003596e0             14000003                JMP 3(PC)                                                                               
  client.go:3338        0x1003596e4             aa1f03f0                MOVD ZR, R16                  
..
```

File: server/stream.go
Comment (kozlovic) [reply server/stream.go:3257]: Actually, out of 6 calls, 4 pass an mt object, so I decided not to. Let me know if you still want me to add a "processJetStreamMsgWithTrace()` or something like that.

**Comment by kozlovic** [comment]
@derekcollison I did a rebase from main (hopefully to not have the route compression flapper issue that I have fixed last week) and made changes based on your comments. I will work now on the services/stream fix and will submit as a separate commit.

File: server/stream.go
Comment (derekcollison) [reply server/stream.go:3257]: Your call, was a suggestion.

**Comment by kozlovic** [comment]
@derekcollison The last commit contains the changes for service/streams. The PR is ready for review.

**Comment by ripienaar** [comment]
@kozlovic to me with the direction of this seems right now and it works as intended.

I do see that one of my servers is not publishing the traces like 50% of the time though and it was stable before the rebase, I've restarted it and so forth but no luck.  Bit vague I am sorry but maybe you have some insights, I can try running it in debug/trace if I must to help get some useful information

**Comment by kozlovic** [comment]
@ripienaar I would need a bit more information indeed. Is it about services or something different? What traces are missing you think? the trace from a server that had egresses to the service provider? (assuming this is about services). Have you run a sub on the trace subject outside of the library to make sure that the trace is really not received as opposed to possibly an issue there?

Also, could you clarify if it was working fine before the rebase or before the changes made in the last commit, specifically for services/streams? If not too difficult to setup, could you build server from before-last commit and change back your configuration to "share:true" and see if that works fine then, which then would mean it is in the latest commit, or if the issue is due to a rebase and not really related to the most recent commits?

I am running the service tests with -count to see if I can reproduce an issue with existing tests. Any information that you could provide will help. Thanks!

**Comment by ripienaar** [comment]
In this case a working path:

```
Client "tracer" cid:1372 cluster:"lon" server:"n3-lon" version:"2.11.0-tracing4"
    Service Import from:"weather.lon" to:"weather.lon" account:"weather"
==> Gateway "n1-nyc" gid:729
    Gateway "n3-lon" gid:665 cluster:"nyc" server:"n1-nyc" version:"2.11.0-tracing4"
    --| Client "NATS CLI Version development" cid:846 subject:"weather.>" queue:"NATS-RPLY-22"
```

The failing path ends at the gateway connection to n1-nyc, n1-nyc isnt sending back traces regardless of which gid it entered on.  Everything else is fine. Above happens to be a import but same happens in the same account.

I'll run in trace and send details, but sounding a bit like a zoom might be best :)

**Comment by kozlovic** [comment]
@ripienaar I would like a zoom, send me an invite to my personal email.

**Comment by kozlovic** [comment]
Are you using a short-lived subscription to receive the traces? If so, could it be a subscription interest propagation/race that causes trace message to not be received back from the GW? Are you using "trace only" or with message delivered? Do you see the message actually delivered in the GW, but just the trace not coming back?

I would run a subscription on the trace subject (if using wildcard, say trace.* while your tracer may create a sub on trace.1, etc...) and have that subscription run prior to send the trace and see if it is then received. Same maybe having a subscription on the GW server(s) on the trace subject to see if a trace is emitted at least over there.

**Comment by ripienaar** [comment]
> Are you using a short-lived subscription to receive the traces? If so, could it be a subscription interest propagation/race that causes trace message to not be received back from the GW? Are you using "trace only" or with message delivered? Do you see the message actually delivered in the GW, but just the trace not coming back?
> 
> I would run a subscription on the trace subject (if using wildcard, say trace.* while your tracer may create a sub on trace.1, etc...) and have that subscription run prior to send the trace and see if it is then received. Same maybe having a subscription on the GW server(s) on the trace subject to see if a trace is emitted at least over there.

It does appear to be time related yeah - I added the flush below before the publish which was enough before, I guess now my network conditions has changed a bit and its more affected.

As it is with the `time.Sleep(time.Second)` its reliable, hmm.

```go
	sub, err := nc.SubscribeSync(nc.NewRespInbox())
	if err != nil {
		return nil, err
	}
	defer sub.Unsubscribe()

	err = nc.Flush()
	if err != nil {
		return nil, err
	}

	time.Sleep(time.Second)

	newMsg := nats.NewMsg(traceMsg.Subject)
	newMsg.Header = traceMsg.Header
	copy(newMsg.Data, traceMsg.Data)
	newMsg.Header.Set("Accept-Encoding", "snappy")
	newMsg.Header.Set(server.MsgTraceSendTo, sub.Subject)
	if !deliverToDest {
		newMsg.Header.Set(server.MsgTraceOnly, "true")
	}

	err = nc.PublishMsg(newMsg)
	if err != nil {
		return nil, err
	}

	return GetMsgTrace(sub, sub.Subject, timeout, trace)
```

**Comment by kozlovic** [comment]
Yes, because remember that the code in GW to avoid sub interest race were more about req-reply, so the check is done if a message has a reply subject and the subscription was made a short while ago. But here, you don't have a reply subject, and rightfully so since this is not a request reply.
So I would recommend that the tool create a wildcard subscription for traces that is created on startup and then each "trace" request is using a unique id as the last token. The setup of that subscription at the start of the app (including flush) may be enough to let the subscription interest propagate.

**Comment by kozlovic** [comment]
> I'll just add a log line and sleep :)

Not ideal, but this is a more general issue of interest propagation with gateways, unfortunately.

**Comment by ripienaar** [comment]
thanks @kozlovic

**Comment by ripienaar** [comment]
I am good to merge this @derekcollison @kozlovic @bruth

**Comment by derekcollison** [comment]
ok so ready for final review?

Will try to get to this tomorrow.

**Comment by kozlovic** [comment]
@derekcollison Actually, could you let me check if I could replace addition of boolean in server's INFO/CONNECT in order to know if the remote supports message tracing, with the use of the `Proto` field. I will update the PR with a note when I am done (either with a new commit or saying that I am not changing and it is ok to review).

**Comment by kozlovic** [comment]
@derekcollison Ok, I have made the changes. So you are good to go to review the last 2 commits, which were added since you have last reviewed. Thanks!

File: server/accounts.go
Comment (derekcollison) [inline server/accounts.go:2511]: Maybe coordinate with @aricart to get it added.

File: server/accounts.go
Comment (derekcollison) [inline server/accounts.go:131]: Update copyright to 2024?

**Review by derekcollison — APPROVED** [review]
LGTM - two minor comments.

File: server/accounts.go
Comment (kozlovic) [reply server/accounts.go:2511]: There would be also some code to add in updateAccountClaimsWithRefresh() to possibly call the function to set the service export allow trace property. And I would like to add a test with JWT of course. I would prefer to wait and add the JWT support before merging, but that would depend on how long it would take to get the JWT library updated. It can of course all be added after the merge. Your call Derek.
@aricart Would you mind adding a AllowTrace boolean property to a ServiceExport and StreamImport? If you are using common structures, validation should make it an error if this property is set on a ServiceImport or a StreamExport (the opposite of where it is valid).

File: server/accounts.go
Comment (kozlovic) [reply server/accounts.go:131]: Done

**Comment by MikaelElkiaer** [comment]
After looking through this PR and related ones, it is still unclear to me whether spans can be exported.

Based on [this reply](https://github.com/nats-io/nats-server/pull/5014#issuecomment-1918669354) I guess it is not yet possible to export OTEL spans to a service such as OTEL Collector or Grafana Tempo.
If not, is this in the pipeline? I would love to export spans to OTEL Collector to link NATS spans to my application spans in case of erroneous requests.

**Comment by ripienaar** [comment]
We do not export spans exactly - we have no otel dependencies etc - but a receiver for these messages can be written that turns them into spans and sends them to a collector.

We have some early reviewers who have written such receivers and reported success with that approach

**Comment by MikaelElkiaer** [comment]
> We do not export spans exactly - we have no otel dependencies etc - but a receiver for these messages can be written that turns them into spans and sends them to a collector.
> 
> We have some early reviewers who have written such receivers and reported success with that approach

Thanks for clarifying.
I had assumed it was something written to stdout at first, but it makes sense that it is a message subject.
Does it need to be enabled? Is there any docs on this?

**Comment by ripienaar** [comment]
It is available in the preview releases and main of the CLI

see https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-41.md

**Pull request merged**
